File: typescript/src/data-structures/stack.ts
/**
 * Stack implementation using a linked list.
 * Time: O(1) for all operations
 * Space: O(n)
 */

interface StackNode<T> {
  value: T;
  next: StackNode<T> | null;
}

export class Stack<T> {
  private head: StackNode<T> | null = null;
  private _size: number = 0;

  push(value: T): void {
    const node: StackNode<T> = { value, next: this.head };
    this.head = node;
    this._size++;
  }

  pop(): T | undefined {
    if (!this.head) {
      return undefined;
    }
    const value = this.head.value;
    this.head = this.head.next;
    this._size--;
    return value;
  }

  peek(): T | undefined {
    return this.head?.value;
  }

  isEmpty(): boolean {
    return this._size === 0;
  }

  size(): number {
    return this._size;
  }

  clear(): void {
    this.head = null;
    this._size = 0;
  }
}

File: zig/src/data_structures/stack.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// LIFO stack implementation.
/// Time: O(1) for all operations
/// Space: O(n)
pub fn Stack(comptime T: type) type {
    return struct {
        const Self = @This();

        items: std.ArrayList(T),

        pub fn init(allocator: Allocator) Self {
            return .{
                .items = std.ArrayList(T).init(allocator),
            };
        }

        pub fn deinit(self: *Self) void {
            self.items.deinit();
        }

        pub fn push(self: *Self, value: T) !void {
            try self.items.append(value);
        }

        pub fn pop(self: *Self) ?T {
            return self.items.popOrNull();
        }

        pub fn peek(self: *const Self) ?T {
            if (self.items.items.len == 0) return null;
            return self.items.items[self.items.items.len - 1];
        }

        pub fn isEmpty(self: *const Self) bool {
            return self.items.items.len == 0;
        }

        pub fn size(self: *const Self) usize {
            return self.items.items.len;
        }

        pub fn clear(self: *Self) void {
            self.items.clearRetainingCapacity();
        }
    };
}

test "Stack operations" {
    const allocator = std.testing.allocator;
    var stack = Stack(i32).init(allocator);
    defer stack.deinit();

    try stack.push(1);
    try stack.push(2);
    try stack.push(3);

    try std.testing.expectEqual(@as(?i32, 3), stack.pop());
    try std.testing.expectEqual(@as(?i32, 2), stack.pop());
    try std.testing.expectEqual(@as(?i32, 1), stack.pop());
    try std.testing.expectEqual(@as(?i32, null), stack.pop());
}
